verilog/tools/kythe/indexing_facts_tree_extractor.cc

```diff
@@ -263,5 +277,93 @@ void IndexingFactsTreeExtractor::ExtractNetDeclaration(
   }
 }
 
+void IndexingFactsTreeExtractor::ExtractFunction(
+    const verible::SyntaxTreeNode& function_declaration_node) {
+  IndexingNodeData function_node_data(IndexingFactType::kFunctionOrTask);
+  IndexingFactNode function_node(function_node_data);
+
+  // Extract function name.
+  const auto& function_name_token =
+      GetFunctionNameTokenInfo(function_declaration_node);
+  const Anchor function_name_anchor(function_name_token, context_.base);
+  function_node.Value().AppendAnchor(function_name_anchor);
+
+  {
+    const IndexingFactsTreeContext::AutoPop p(&facts_tree_context_,
+                                              &function_node);
+    // Extract function ports.
+    ExtractFunctionTaskPort(function_declaration_node);
+
+    // Extract function body.
+    const verible::SyntaxTreeNode& function_body =
+        GetFunctionBlockStatmentList(function_declaration_node);
+    Visit(function_body);
+  }
+
+  facts_tree_context_.top().NewChild(function_node);
+}
+
+void IndexingFactsTreeExtractor::ExtractTask(
+    const verible::SyntaxTreeNode& task_declaration_node) {
+  IndexingNodeData task_node_data(IndexingFactType::kFunctionOrTask);
+  IndexingFactNode task_node(task_node_data);
+
+  // Extract task name.
+  const auto& task_name_token = GetTaskNameTokenInfo(task_declaration_node);
+  const Anchor task_name_anchor(task_name_token, context_.base);
+  task_node.Value().AppendAnchor(task_name_anchor);
+
+  {
+    const IndexingFactsTreeContext::AutoPop p(&facts_tree_context_, &task_node);
+
+    // Extract task ports.
+    ExtractFunctionTaskPort(task_declaration_node);
+
+    // Extract task body.
+    const verible::SyntaxTreeNode& task_body =
+        GetTaskStatmentList(task_declaration_node);
+    Visit(task_body);
+  }
+
+  facts_tree_context_.top().NewChild(task_node);
+}
+
+void IndexingFactsTreeExtractor::ExtractFunctionTaskPort(
+    const verible::SyntaxTreeNode& function_declaration_node) {
+  const std::vector<verible::TreeSearchMatch> ports =
+      FindAllTaskFunctionPortDeclarations(function_declaration_node);
+
+  for (const verible::TreeSearchMatch port : ports) {
+    const verible::SyntaxTreeLeaf* leaf =
+        GetIdentifierFromTaskFunctionPortItem(*port.match);
+
+    facts_tree_context_.top().NewChild(
+        IndexingNodeData({Anchor(leaf->get(), context_.base)},
+                         IndexingFactType::kVariableDefinition));
```

Comment:
Is there a different fact type for variables that come through ports or parameters?

Reply:
Yes, in Kythe function can have `param` facts.
For now I am not using `param` facts so I decided to treat the parameters as variables.

Reply:
The `param` fact is an extra fact it won't remove the variable fact.

Reply:
Down the road, I do think the distinction between port/parameter and local variable would be helpful.  Can you leave a //TODO?

Reply:
This will most likely change after merging the PRs but I left a TODO for now.